apps/emqx_management/test/emqx_mgmt_api_clients_SUITE.erl

```diff
@@ -171,6 +200,257 @@ t_clients(_) ->
+                    list_request(APIPort, "node=" ++ atom_to_list(N1))
+                )
+            ),
+            ok = emqtt:stop(C8),
+            ok = erpc:call(N1, emqx_persistent_session_ds, destroy_session, [ClientId4]),
+
+            %% Pagination with mixed clients
+            ClientId5 = <<"c5">>,
+            ClientId6 = <<"c6">>,
+            ClientId7 = <<"c7">>,
+            ClientId8 = <<"c8">>,
+            %% persistent
+            C9 = Connect(#{port => Port1, clientid => ClientId5}),
+            C10 = Connect(#{port => Port2, clientid => ClientId6}),
+            %% in-memory
+            C11 = Connect(#{port => Port1, clientid => ClientId7, expiry => 0, clean_start => true}),
+            C12 = Connect(#{port => Port2, clientid => ClientId8, expiry => 0, clean_start => true}),
+
+            P1 = list_request(APIPort, "limit=3&page=1"),
+            P2 = list_request(APIPort, "limit=3&page=2"),
+            ?assertMatch(
+                {ok,
+                    {{_, 200, _}, _, #{
+                        <<"data">> := [_, _, _],
+                        <<"meta">> := #{
+                            <<"count">> := 4,
+                            <<"hasnext">> := true
+                        }
+                    }}},
+                P1
+            ),
+            ?assertMatch(
+                {ok,
+                    {{_, 200, _}, _, #{
+                        <<"data">> := [_],
+                        <<"meta">> := #{
+                            <<"count">> := 4,
+                            <<"hasnext">> := false
+                        }
+                    }}},
+                P2
+            ),
+            {ok, {_, _, #{<<"data">> := R1}}} = P1,
+            {ok, {_, _, #{<<"data">> := R2}}} = P2,
+            ?assertEqual(
+                lists:sort([ClientId5, ClientId6, ClientId7, ClientId8]),
+                lists:sort(lists:map(fun(#{<<"clientid">> := CId}) -> CId end, R1 ++ R2))
+            ),
+            ?assertMatch(
+                {ok,
+                    {{_, 200, _}, _, #{
+                        <<"data">> := [_, _],
+                        <<"meta">> := #{
+                            <<"count">> := 4,
+                            <<"hasnext">> := true
+                        }
+                    }}},
+                list_request(APIPort, "limit=2&page=1")
+            ),
+            %% Disconnect persistent sessions
+            lists:foreach(fun emqtt:stop/1, [C9, C10]),
+
+            P3 = list_request(APIPort, "limit=3&page=1"),
+            P4 = list_request(APIPort, "limit=3&page=2"),
+            ?assertMatch(
+                {ok,
+                    {{_, 200, _}, _, #{
+                        <<"data">> := [_, _, _],
+                        <<"meta">> := #{
+                            <<"count">> := 4,
+                            <<"hasnext">> := true
+                        }
+                    }}},
+                P1
```

Comment:
Q: Why do we match `P1` here, and not `P3`? Similar question about `P2` below.

Reply:
D'oh, bad copy and paste...  :see_no_evil: